src/main/java/com/hierynomus/smbj/session/Session.java

```diff
@@ -179,10 +182,10 @@ private Share connectTree(final String shareName) {
             TreeConnect treeConnect = new TreeConnect(treeId, smbPath, this, response.getCapabilities(), connection, bus);
 
             Share share;
-            if (response.isDiskShare() && dfsPathResolver != null && response.getCapabilities().contains(SMB2ShareCapabilities.SMB2_SHARE_CAP_DFS)) {
-                share = new DFSDiskShare(smbPath, treeConnect, dfsPathResolver);
-            } else if (response.isDiskShare()) {
-                share = new DiskShare(smbPath, treeConnect);
+/*            if (response.isDiskShare() && pathResolver != null && response.getCapabilities().contains(SMB2ShareCapabilities.SMB2_SHARE_CAP_DFS)) {
```

Comment:
With this change, the library will always attempt to perform DFS referral requests even though the share may not support it. Everything can be solved with an extra level of indirection. Perhaps pathResolver should be a `Function<Set<Smb2ShareCapabilities>, PathResolver>` instead so that selection of the actual implementation can be deferred until this point.
In other words, pass a PathResolver factory to the Session instead of a concrete PathResolver. Depending on the capabilities, the factory would either return the DFSPathResolver decorator or the wrapped SymlinkPathResolver directly.

Reply:
That being said DFSPathResolver only takes specific actions when PATH_NOT_COVERED is returned. You're unlikely to get that error unless DFS is enabled to begin with. Does the spec say anything about dealing with PATH_NOT_COVERED when SMB2_SHARE_CAP_DFS is not set?